File: sdks/python/workersdk/worker.py
import json
import time
import uuid
import psycopg2
from psycopg2.extras import DictCursor
from typing import Callable, Dict

class Worker:
    def __init__(self):
        self.handlers: Dict[str, Callable[[str], str]] = {}
        self.connection = None
        self.worker_id = f"python-worker-{str(uuid.uuid4())[:8]}"

    def register(self, action: str, handler: Callable[[str], str]):
        self.handlers[action] = handler

    def start(self, dsn: str = "postgresql://admin:admin@postgres:5432/wf_engine"):
        self._connect_with_retry(dsn)
        
        print(f"Worker {self.worker_id} started, listening for jobs via Postgres SKIP LOCKED...")
        
        try:
            while True:
                self._poll_and_process()
                time.sleep(0.5)  # Poll every 500ms
        except KeyboardInterrupt:
            print("Worker stopping...")
        finally:
            if self.connection:
                self.connection.close()

    def _poll_and_process(self):
        if not self.handlers:
            return

        actions = list(self.handlers.keys())
        
        with self.connection.cursor(cursor_factory=DictCursor) as cursor:
            try:
                claim_sql = """
                    UPDATE workflow_job 
                    SET status = 'PROCESSING', locked_by = %s
                    WHERE id = (
                        SELECT id FROM workflow_job 
                        WHERE status = 'PENDING' AND action = ANY(%s)
                        ORDER BY created_at ASC
                        LIMIT 1 
                        FOR UPDATE SKIP LOCKED
                    )
                    RETURNING id, workflow_run_id, action, payload
                """
                
                cursor.execute(claim_sql, (self.worker_id, actions))
                job_row = cursor.fetchone()

                if not job_row:
                    self.connection.commit()
                    return # No jobs available

                job_id = job_row['id']
                run_id = job_row['workflow_run_id']
                action = job_row['action']
                payload = job_row['payload']

                print(f"Processing job: {action} for run: {run_id}")

                handler = self.handlers.get(action)
                status = "SUCCESS"
                
                try:
                    result = handler(payload)
                except Exception as e:
                    # Safely construct JSON error
                    result = json.dumps({"error": str(e)})
                    status = "FAILED"

                update_sql = """
                    UPDATE workflow_job 
                    SET status = 'COMPLETED', result = %s, result_status = %s
                    WHERE id = %s
                """
                cursor.execute(update_sql, (result, status, job_id))
                
                self.connection.commit()
                
            except Exception as e:
                self.connection.rollback()
                print(f"Error in polling transaction: {e}")

    def _connect_with_retry(self, dsn: str):
        for attempt in range(30):
            try:
                self.connection = psycopg2.connect(dsn)
                # Test connection
                with self.connection.cursor() as cur:
                    cur.execute('SELECT 1')
                return
            except Exception as e:
                print(f"Postgres not ready, retrying ({attempt + 1}/30)...")
                time.sleep(2)
        else:
            raise Exception("Could not connect to Postgres after 30 attempts")

File: sdks/node/src/worker.ts
import { Client } from 'pg';
import { randomUUID } from 'crypto';

export class Worker {
    private handlers: Map<string, (payload: string) => any>;
    private client: Client | null;
    private workerId: string;

    constructor() {
        this.handlers = new Map();
        this.client = null;
        this.workerId = 'node-worker-' + randomUUID().substring(0, 8);
    };

    register(action: string, handler: (payload: string) => any) {
        this.handlers.set(action, handler);
    };

    async start(connectionString: string = "postgres://admin:admin@postgres:5432/wf_engine"): Promise<void> {
        
        await this.connectWithRetry(connectionString);

        console.log(`Worker ${this.workerId} started, listening for jobs via Postgres SKIP LOCKED...`);

        // Start the polling loop
        const poll = async () => {
            try {
                await this.pollAndProcess();
            } catch (e) {
                console.error("Error in polling loop:", e);
            }
            setTimeout(poll, 500); // Poll every 500ms
        };

        await poll();
    };

    private async pollAndProcess(): Promise<void> {
        if (this.handlers.size === 0 || !this.client) {
            return;
        }

        const actions = Array.from(this.handlers.keys());

        const placeholders = actions.map((_, i) => `$${i + 1}`).join(', ');

        try {
            await this.client.query('BEGIN');

            const claimSql = `
                UPDATE workflow_job 
                SET status = 'PROCESSING', locked_by = '${this.workerId}'
                WHERE id = (
                    SELECT id FROM workflow_job 
                    WHERE status = 'PENDING' AND action IN (${placeholders})
                    ORDER BY created_at ASC
                    LIMIT 1 
                    FOR UPDATE SKIP LOCKED
                )
                RETURNING id, workflow_run_id as "workflowRunId", action, payload
            `;

            const claimResult = await this.client.query(claimSql, actions);

            if (claimResult.rows.length === 0) {
                await this.client.query('COMMIT');
                return; // No jobs available
            }

            const job = claimResult.rows[0];
            console.log(`Processing job: ${job.action} for run: ${job.workflowRunId}`);

            const handler = this.handlers.get(job.action);
            let resultOutput: string;
            let status: string = "SUCCESS";

            try {
                // Support both sync and async handlers
                const result = await handler!(job.payload);
                resultOutput = typeof result === 'string' ? result : JSON.stringify(result);
            } catch (e) {
                status = "FAILED";
                resultOutput = JSON.stringify({ error: String(e) });
            }

            const updateSql = `
                UPDATE workflow_job 
                SET status = 'COMPLETED', result = $1, result_status = $2
                WHERE id = $3
            `;
            
            await this.client.query(updateSql, [resultOutput, status, job.id]);

            await this.client.query('COMMIT');

        } catch (e) {
            await this.client.query('ROLLBACK');
            throw e;
        }
    }

    private async connectWithRetry(connectionString: string): Promise<void> {
        for (let attempt = 1; attempt <= 30; attempt++) {
            try {
                this.client = new Client({ connectionString });
                await this.client.connect();
                await this.client.query('SELECT 1');
                return;
            } catch (e) {
                console.log(`Postgres not ready, retrying (${attempt}/30)...`);
                await new Promise(resolve => setTimeout(resolve, 2000));
                if (attempt === 30) {
                    throw new Error('Could not connect to Postgres after 30 attempts');
                }
            }
        }
    }
}
